src/drivers/gnss/septentrio/septentrio.cpp

```diff
@@ -0,0 +1,1439 @@
+				// other data, but it's really large: >800B)
+				_p_report_sat_info->timestamp = hrt_absolute_time();
+				_p_report_sat_info->count = _report_gps_pos.satellites_used;
+				ret = 2;
+			}
+
+		} else {
+			_report_gps_pos.satellites_used = 0;
+		}
+
+		_report_gps_pos.latitude_deg = _buf.payload_pvt_geodetic.latitude * M_RAD_TO_DEG;
+		_report_gps_pos.longitude_deg = _buf.payload_pvt_geodetic.longitude * M_RAD_TO_DEG;
+		_report_gps_pos.altitude_ellipsoid_m = _buf.payload_pvt_geodetic.height;
+		_report_gps_pos.altitude_msl_m = _buf.payload_pvt_geodetic.height - static_cast<double>
+						(_buf.payload_pvt_geodetic.undulation);
+
+		/* H and V accuracy are reported in 2DRMS, but based off the uBlox reporting we expect RMS.
+		 * Devide by 100 from cm to m and in addition divide by 2 to get RMS. */
+		_report_gps_pos.eph = static_cast<float>(_buf.payload_pvt_geodetic.h_accuracy) / 200.0f;
+		_report_gps_pos.epv = static_cast<float>(_buf.payload_pvt_geodetic.v_accuracy) / 200.0f;
+
+		_report_gps_pos.vel_n_m_s = static_cast<float>(_buf.payload_pvt_geodetic.vn);
+		_report_gps_pos.vel_e_m_s = static_cast<float>(_buf.payload_pvt_geodetic.ve);
+		_report_gps_pos.vel_d_m_s = -1.0f * static_cast<float>(_buf.payload_pvt_geodetic.vu);
+		_report_gps_pos.vel_m_s = sqrtf(_report_gps_pos.vel_n_m_s * _report_gps_pos.vel_n_m_s +
+					       _report_gps_pos.vel_e_m_s * _report_gps_pos.vel_e_m_s);
+
+		_report_gps_pos.cog_rad = static_cast<float>(_buf.payload_pvt_geodetic.cog) * M_DEG_TO_RAD_F;
+		_report_gps_pos.c_variance_rad = 1.0f * M_DEG_TO_RAD_F;
+
+		// _buf.payload_pvt_geodetic.cog is set to -2*10^10 for velocities below 0.1m/s
+		if (_buf.payload_pvt_geodetic.cog > 360.0f) {
+			_buf.payload_pvt_geodetic.cog = NAN;
+		}
+
+		_report_gps_pos.time_utc_usec = 0;
+#ifndef NO_MKTIME
+		struct tm timeinfo;
+		time_t epoch;
+
+		// Convert to unix timestamp
+		memset(&timeinfo, 0, sizeof(timeinfo));
+
+		timeinfo.tm_year = 1980 - 1900;
+		timeinfo.tm_mon = 0;
+		timeinfo.tm_mday = 6 + _buf.WNc * 7;
+		timeinfo.tm_hour = 0;
+		timeinfo.tm_min = 0;
+		timeinfo.tm_sec = _buf.TOW / 1000;
+
+		epoch = mktime(&timeinfo);
+
+		if (epoch > GPS_EPOCH_SECS) {
+			// FMUv2+ boards have a hardware RTC, but GPS helps us to configure it
+			// and control its drift. Since we rely on the HRT for our monotonic
+			// clock, updating it from time to time is safe.
+
+			timespec ts;
+			memset(&ts, 0, sizeof(ts));
+			ts.tv_sec = epoch;
+			ts.tv_nsec = (_buf.TOW % 1000) * 1000 * 1000;
+			set_clock(ts);
+
+			_report_gps_pos.time_utc_usec = static_cast<uint64_t>(epoch) * 1000000ULL;
+			_report_gps_pos.time_utc_usec += (_buf.TOW % 1000) * 1000;
+		}
+
+#endif
+		_report_gps_pos.timestamp = hrt_absolute_time();
+		_last_timestamp_time = _report_gps_pos.timestamp;
+		_rate_count_vel++;
+		_rate_count_lat_lon++;
+		// NOTE: Isn't this just `ret |= (_msg_status == 7)`?
+		ret |= (_msg_status == 7) ? 1 : 0;
```

Comment:
Just copied this, but it is written in a confusing way. I don't want to touch this code without asking as I'm not 100% certain how the parsing implementation works (the purpose of some variables isn't immediately clear and there is no documentation for them).